### PR TITLE
Unsafe code replaced by ReadOnlyMemory<> equivalent

### PR DESCRIPTION
**Pull request by YohDeadfall**
The fix removes unsafe code and replaces it by safe version using `ReadOnlyMemory<char>` and `Span<T>`. There is a small performance regression caused by the change:

``` ini

BenchmarkDotNet=v0.12.1, OS=Windows 10.0.20226
Intel Xeon CPU E5-2687W 0 3.10GHz, 2 CPU, 32 logical and 16 physical cores
.NET Core SDK=5.0.100-rc.1.20452.10
  [Host]     : .NET Core 3.1.8 (CoreCLR 4.700.20.41105, CoreFX 4.700.20.41903), X64 RyuJIT
  Job-JEAYIC : .NET Core 5.0.0 (CoreCLR 5.0.20.45114, CoreFX 5.0.20.45114), X64 RyuJIT

Runtime=.NET Core 5.0  

```

|      Method |   value |     Mean |    Error |   StdDev | Code Size |
|------------ |-------- |---------:|---------:|---------:|----------:|
| **GetItemUnsafe** | **Char[1]** | **30.80 ns** | **0.079 ns** | **0.074 ns** |     **156 B** |
| GetItemSafe | Char[1] | 35.94 ns | 0.036 ns | 0.032 ns |     193 B |
| **GetItemUnsafe** | **Char[2]** | **28.88 ns** | **0.493 ns** | **0.437 ns** |     **156 B** |
| GetItemSafe | Char[2] | 36.13 ns | 0.166 ns | 0.147 ns |     193 B |
| **GetItemUnsafe** | **Char[3]** | **29.70 ns** | **0.320 ns** | **0.268 ns** |     **156 B** |
| GetItemSafe | Char[3] | 34.01 ns | 0.016 ns | 0.015 ns |     193 B |

```csharp
private readonly Dictionary<StringOrCharArray, int> _strings = new Dictionary<StringOrCharArray, int>();

public Program()
{
    for (var i = 1; i < 10; i++)
    for (var c = 'a'; c <= 'z'; c++)
        _strings.Add(new string(c, i), i);
}

[Benchmark]
[ArgumentsSource(nameof(GetCharArrays))]
public int GetItem(char[] value) =>
    _strings[new StringOrCharArray(value, 0, value.Length)];

public IEnumerable<string> GetStrings()
{
    yield return "a";
    yield return "bb";
    yield return "ccc";
}

public IEnumerable<char[]> GetCharArrays()
{
    foreach (var s in GetStrings())
        yield return s.ToCharArray();
}
```

### CODE REVIEW

**Comment by huoyaoyuan** [comment]
It looks better to remove this type and use `ReadOnlyMemory<char>` directly.

**Comment by YohDeadfall** [comment]
Not possible because of the difference in implementations of `Equals` and `GetHashCode` methods. `ReadOnlyMemory<T>` doesn't take the content into account while this one does.

File: src/libraries/Common/src/System/Text/StringOrCharArray.cs [hunk 1]
```diff
@@ -13,133 +13,26 @@ namespace System.Text
     /// </summary>
     internal readonly struct StringOrCharArray : IEquatable<StringOrCharArray>
```
Comment (stephentoub) [inline src/libraries/Common/src/System/Text/StringOrCharArray.cs:14]: This name and the comment above it are no longer accurate.

**Comment by stephentoub** [comment]
> Not possible because of the difference in implementations of Equals and GetHashCode methods. ReadOnlyMemory<T> doesn't take the content into account while this one does.

How about just having a special comparer for `ReadOnlyMemory<char>` rather than having this wrapper type?

**Comment by YohDeadfall** [comment]
Thought about the comparer, but it should be instantiated and used as an interface while the current implementation allows to inline methods more aggressively since it's a struct and there is an optimization for default equality comparers.

**Comment by YohDeadfall** [comment]
Well, in that case will remove the type and add a comparer directly to `ConsolePal` where it's only used.

**Comment by YohDeadfall** [comment]
> Not sure what you mean about instantiated as an interface, though. Can you elaborate? We do want to avoid lots of unnecessary allocation, and at quick glance I'm not seeing why an interface would be needed.

I meant that there will be an instance of a comparer which members will be accessed via virtual calls, but as you said that's not a problem.

**Comment by stephentoub** [comment]
Cool

File: src/libraries/System.Console/src/System/ConsolePal.Unix.cs [hunk 2]
```diff
@@ -1189,8 +1190,8 @@ private void AddKey(TermInfo.Database db, TermInfo.WellKnownStrings keyId, Conso
 
             private void AddKey(TermInfo.Database db, TermInfo.WellKnownStrings keyId, ConsoleKey key, bool shift, bool alt, bool control)
             {
-                string? keyFormat = db.GetString(keyId);
-                if (!string.IsNullOrEmpty(keyFormat))
+                var keyFormat = db.GetString(keyId).AsMemory();
```
Comment (eiriktsarpalis) [inline src/libraries/System.Console/src/System/ConsolePal.Unix.cs:1193]:
```suggestion
                ReadOnlyMemory<char> keyFormat = db.GetString(keyId).AsMemory();
```

per the [coding guidelines](https://github.com/dotnet/runtime/blob/master/docs/coding-guidelines/coding-style.md).

File: src/libraries/System.Console/src/System/ConsolePal.Unix.cs [hunk 3]
```diff
@@ -1205,8 +1206,8 @@ private void AddPrefixKey(TermInfo.Database db, string extendedNamePrefix, Conso
 
             private void AddKey(TermInfo.Database db, string extendedName, ConsoleKey key, bool shift, bool alt, bool control)
             {
-                string? keyFormat = db.GetExtendedString(extendedName);
-                if (!string.IsNullOrEmpty(keyFormat))
+                var keyFormat = db.GetExtendedString(extendedName).AsMemory();
```
Comment (eiriktsarpalis) [inline src/libraries/System.Console/src/System/ConsolePal.Unix.cs:1209]:
```suggestion
                ReadOnlyMemory<char> keyFormat = db.GetExtendedString(extendedName).AsMemory();
```

File: src/libraries/System.Console/src/System/ConsolePal.Unix.cs [hunk 4]
```diff
@@ -1513,5 +1514,36 @@ private static void HandleBreakEvent(object? state)
...
         }
+
+        public sealed class ReadOnlyMemoryContentComparer : IEqualityComparer<ReadOnlyMemory<char>>
+        {
+            public bool Equals(ReadOnlyMemory<char> x, ReadOnlyMemory<char> y) =>
+                x.Span.SequenceEqual(y.Span);
+
+            public int GetHashCode(ReadOnlyMemory<char> obj)
+            {
+                // This hash code is a simplified version of some of the code in String,
```
Comment (eiriktsarpalis) [inline src/libraries/System.Console/src/System/ConsolePal.Unix.cs:1525]: @stephentoub we use [Marvin.cs](https://github.com/dotnet/runtime/blob/master/src/libraries/Common/src/System/Marvin.cs) for `ReadOnlyMemory<byte>` hashcode computation. Wondering if a similar (non-randomized) implementation exists for char buffers and if not, would it make sense to move this to the common folder?

File: src/libraries/System.Console/src/System/ConsolePal.Unix.cs [hunk 5]
```diff
@@ -1513,5 +1514,36 @@ private static void HandleBreakEvent(object? state)
...
+                // when not using randomized hash codes.  We don't use string's GetHashCode
+                // because we need to be able to use the exact same algorithms on a char[].
+                // As such, this should not be used anywhere there are concerns around
+                // hash-based attacks that would require a better code.
+
+                int hash1 = (5381 << 16) + 5381;
+                int hash2 = hash1;
+                var span = obj.Span;
```
Comment (eiriktsarpalis) [inline src/libraries/System.Console/src/System/ConsolePal.Unix.cs:1533]:
```suggestion
                ReadOnlySpan<char> span = obj.Span;
```

**Review by eiriktsarpalis — APPROVED** [review]
Other than a few minor nits, this LGTM. Thank you for the contribution!

File: src/libraries/System.Console/src/System/ConsolePal.Unix.cs [hunk 6]
```diff
@@ -1513,5 +1514,36 @@ private static void HandleBreakEvent(object? state)
                 }
             }
         }
+
+        public sealed class ReadOnlyMemoryContentComparer : IEqualityComparer<ReadOnlyMemory<char>>
```
Comment (stephentoub) [inline src/libraries/System.Console/src/System/ConsolePal.Unix.cs:1518]:
```suggestion
        private sealed class ReadOnlyMemoryContentComparer : IEqualityComparer<ReadOnlyMemory<char>>
```

File: src/libraries/System.Console/src/System/ConsolePal.Unix.cs [hunk 7]
```diff
@@ -1513,5 +1514,36 @@ private static void HandleBreakEvent(object? state)
...
+        {
+            public bool Equals(ReadOnlyMemory<char> x, ReadOnlyMemory<char> y) =>
+                x.Span.SequenceEqual(y.Span);
+
+            public int GetHashCode(ReadOnlyMemory<char> obj)
```
Comment (stephentoub) [inline src/libraries/System.Console/src/System/ConsolePal.Unix.cs:1523]: This whole implementation can just be:
```C#
public int GetHashCode(ReadOnlyMemory<char> obj) => string.GetHashCode(obj.Span);
```

File: src/libraries/System.Console/src/System/ConsolePal.Unix.cs (hunk 6 above)
Comment (YohDeadfall) [reply src/libraries/System.Console/src/System/ConsolePal.Unix.cs:1518]: Whoops, forgot to change the modifier when copied it from a benchmark project.

File: src/libraries/System.Console/src/System/ConsolePal.Unix.cs (hunk 4 above)
Comment (stephentoub) [reply src/libraries/System.Console/src/System/ConsolePal.Unix.cs:1525]: See the comment I just added above. ;-)

File: src/libraries/System.Console/src/System/ConsolePal.Unix.cs [hunk 8]
```diff
@@ -1024,7 +1024,8 @@ internal class TerminalFormatStrings
             /// The dictionary of keystring to ConsoleKeyInfo.
             /// Only some members of the ConsoleKeyInfo are used; in particular, the actual char is ignored.
             /// </summary>
-            public readonly Dictionary<StringOrCharArray, ConsoleKeyInfo> KeyFormatToConsoleKey = new Dictionary<StringOrCharArray, ConsoleKeyInfo>();
+            public readonly Dictionary<ReadOnlyMemory<char>, ConsoleKeyInfo> KeyFormatToConsoleKey =
+                new Dictionary<ReadOnlyMemory<char>, ConsoleKeyInfo>(new ReadOnlyMemoryContentComparer());
```
Comment (stephentoub) [inline src/libraries/System.Console/src/System/ConsolePal.Unix.cs:1028]:
```suggestion
                new Dictionary<ReadOnlyMemory<char>, ConsoleKeyInfo>(new ReadOnlyMemoryContentComparer());
                
```

File: src/libraries/System.Console/src/System/ConsolePal.Unix.cs (hunk 4 above)
Comment (stephentoub) [reply src/libraries/System.Console/src/System/ConsolePal.Unix.cs:1525]: (That said, I would not be surprised if we find other places we want to use a ReadOnlyMemoryComparer like this, or even make it public.  If/when that happens, we can certainly move this.  Might make sense now to at least put it in its own file.)

**Review by stephentoub — APPROVED** [review]
Thanks!

**Comment by eiriktsarpalis** [comment]
CI failures are unrelated timeouts on Windows jobs, merging away.

**Pull request merged**